aiida/work/job_calcs.py

```diff
@@ -0,0 +1,226 @@
+"""
+Module containing utilities and classes relating to job calculations running
+on systems that require transport.
+"""
+
+from __future__ import absolute_import
+import contextlib
+from functools import partial
+import time
+from six import iteritems, itervalues
+from tornado import concurrent, gen
+
+from aiida import scheduler as schedulers
+from aiida.common import exceptions
+from . import utils
+
+
+class JobsList(object):
+    """
+    A list of submitted jobs on a machine connected to by transport based on the
+    authorisation information.
+    """
+
+    def __init__(self, authinfo, transport_queue):
+        """
+        :param authinfo: The authinfo used to check the jobs list
+        :type authinfo: :class:`aiida.orm.AuthInfo`
+        :param transport_queue: A transport queue
+        :type: :class:`aiida.work.transports.TransportQueue`
+        """
+        self._authinfo = authinfo
+        self._transport_queue = transport_queue
+        self._loop = transport_queue.loop()
+
+        self._jobs_cache = {}
+        self._last_updated = None  # type: float
+        self._job_update_requests = {}  # Mapping: {job_id: Future}
+        self._update_handle = None
+
+    def get_minimum_update_interval(self):
+        """
+        Get the minimum interval that can be expected between updates of the list
+        :return: The minimum interval
+        :rtype: float
+        """
+        return self._authinfo.computer.get_minimum_job_poll_interval()
+
+    def get_last_updated(self):
+        """
+        Get the timestamp of when the list was last updated as produced by `time.time()`
+
+        :return: The last update point
+        :rtype: float
+        """
+        return self._last_updated
+
+    @gen.coroutine
+    def _get_jobs_from_scheduler(self):
+        """
+        Get the current jobs list from the scheduler
+
+        :return: A dictionary of {job_id: job info}
+        :rtype: dict
+        """
+        with self._transport_queue.request_transport(self._authinfo) as request:
+            transport = yield request
+
+            scheduler = self._authinfo.computer.get_scheduler()
+            scheduler.set_transport(transport)
+
+            kwargs = {'as_dict': True}
+            if scheduler.get_feature('can_query_by_user'):
```

Comment:
What happens if the scheduler cannot query by user? It used to default to just query for a specific job id. I don't see anymore how this is possible.

Reply:
Querying by job id is no longer necessary from this point of the code because it queries all jobs (and caches the results).  In a way that's the point in this class.  It's still useful to query by user in case it's possible just to reduce the number of results.

Reply:
That I understood, but I assumed that given there exists this `scheduler.get_feature('can_query_by_user')` check, certain schedulers *cannot* query by user. If that is in fact the case, this user wide polling will break. So either certain scheduler cannot query on a per user basis and this will break, or all of them can and then this check is superfluous and should be removed

Reply:
Ah, I made an assumption here that if you don't specify a user then it will query from jobs from _all_ users - if that's not true then you're right that this could cause a problem.  Let me check...